url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/net/http/HttpGet.java

```diff
@@ -0,0 +1,270 @@
+import net.hetimatan.net.http.task.client.HttpGetReadBodyTask;
+import net.hetimatan.net.http.task.client.HttpGetReadHeaderTask;
+import net.hetimatan.net.http.task.client.HttpGetRequestTask;
+import net.hetimatan.util.event.EventTask;
+import net.hetimatan.util.event.EventTaskRunner;
+import net.hetimatan.util.event.net.KyoroSocketEventRunner;
+import net.hetimatan.util.event.net.MessageSendTask;
+import net.hetimatan.util.event.net.io.KyoroSocket;
+import net.hetimatan.util.http.HttpGetRequestUri;
+import net.hetimatan.util.http.HttpRequestHeader;
+import net.hetimatan.util.http.HttpResponse;
+import net.hetimatan.util.log.Log;
+
+public class HttpGet {
+
+	public static final String TAG = "HttpGet";
+	public String sId = "[httpget]";
+
+	private HttpGetRequester mCurrentRequest  = null;
+	private HttpGetResponse  mCurrentResponse = null;
+	private KyoroSocket      mCurrentSocket   = null;
+	private String mHost = "127.0.0.1";
+	private String mPath = "/";
+	private int mPort = 80;
+	private EventTaskRunner mRunner = null;
+	private HttpGetTaskManager mTaskManager = new HttpGetTaskManager();
+ 
+	private HttpGetListener mListener = null;
+	public HttpGet() throws IOException {
+		this(null);
+	}
+
+	public HttpGet(HttpGetListener listener) throws IOException {
+		if(listener == null) {
+			mListener = new HttpGetListener.NullHttpGetListener();
+		} else {
+			mListener = listener;
+		}
+	}
+
+	public EventTaskRunner getRunner() {return mRunner;}
+
+	public KyoroSocket getSocket() {return mCurrentSocket;}
+
+	protected HttpGetResponse getGetResponse() {return mCurrentResponse;}
+
+	protected HttpGetRequester createGetRequest() {
+		if (mCurrentRequest == null) {
+			mCurrentRequest = new HttpGetRequester();
+		}
+		return mCurrentRequest;
+	}
+
+	public HttpGet update(String host, String path, int port) throws IOException {
+		mHost = host;
+		mPath = path;
+		mPort = port;
+		sId = "[httpget "+mHost+":"+mPort+mPath+"]";
+		HttpHistory.get().pushMessage(sId+"#update:"+"\n");
+		dispose();
+		return this;
+	}
+
+	public HttpGet update(String location) throws IOException {
+		MarkableFileReader reader = null;
+		try {
+			reader = new MarkableFileReader(location.getBytes());
+			HttpGetRequestUri geturi = HttpGetRequestUri.decode(reader);
+			update(geturi.getHost(), geturi.getMethod(), geturi.getPort());
+		} finally {
+			reader.close();
+		}
+		return this;
+	}
+
+	// <omake>
+	private KyoroFile mBody = null;
+	private boolean mIsPost = false;
+	public void setBody(KyoroFile contents) throws IOException {
+		mBody = contents;
+		mIsPost = true;
+	}
+
+	//</omake>
+	private LinkedHashMap<String,String> mHeaderValue = new LinkedHashMap<>();
+	public void addHeader(String type, String value) {
+		mHeaderValue.put(type, value);
+	}
+	
+	public KyoroSocketEventRunner startTask(KyoroSocketEventRunner runner, EventTask last) {
+		HttpHistory.get().pushMessage(sId+"#startTask"+"\n");
+		mTaskManager.mLast = last;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;last&#39; (index: 1 | type: EventTask). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;runner&#39;. The same parameter type is validated here for example: url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/net/http/HttpServerFront.java:49. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).